### PR TITLE
Hook up all patients in facility view to UI

### PR DESCRIPTION
**Pull request by vinaysshenoy**
Pivotal Story Link: https://www.pivotaltracker.com/story/show/166754208

### CODE REVIEW

**Comment by vinaysshenoy** [comment]
The last part of the feature is pending. But it's a tiny piece because the only thing we need to do is embed this view inside the `AddIdToPatientSearchScreen` and toggle the search button and list on typing. This will be done in another PR.

File: app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityListAdapter.kt [hunk 1]
```diff
@@ -2,19 +2,26 @@ package org.simple.clinic.allpatientsinfacility
...
+
+  private val eventSubject = PublishSubject.create<AllPatientsInFacilityListItem.Event>()
+
+  val listItemEvents: Observable<AllPatientsInFacilityListItem.Event> = eventSubject.hide()
 
   override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderX {
     return ViewHolderX(LayoutInflater.from(parent.context).inflate(viewType, parent, false))
   }
 
   override fun onBindViewHolder(holder: ViewHolderX, position: Int) {
```
Comment (sanchita) [inline app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityListAdapter.kt]: Why is this named `ViewHolderX`?

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityListAdapter.kt]: Just a convention for Kotlin Android Extensions. Package names for kotlin android extensions related code all start with `kotlinx`, and this class is meant to be used for viewholders which need access to synthetic accessors, so `VIewHolderX`.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt [hunk 2]
```diff
@@ -1932,5 +1932,169 @@ class PatientRepositoryAndroidTest {
...
+            facilityB
+        ),
+        recordPatientAtFacility(
+            UUID.fromString("51d7785c-dfe1-4018-8207-e74cd2377f5d"),
+            "Patient F",
+            PatientStatus.DEAD,
+            facilityB
+        ),
+        recordPatientAtFacility(
+            UUID.fromString("072df9ac-6a00-462f-b87e-1a9ca3b24c1b"),
```
Comment (sanchita) [inline app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: We are not using this generated UUID to verify the patient itself.
Why do we need to pass these to `recordPatientAtFacility` and not use the auto-generated UUID in the `TestData` function itself?

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt [hunk 3]
```diff
@@ -1932,5 +1932,169 @@ class PatientRepositoryAndroidTest {
...
+            "Patient D",
```
Comment (sanchita) [inline app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: We are using `fullName` to determine the order.
This is not very obvious reading just the test.
Can you use better names instead like say `OrderByNameA`?
Or it would be a better idea to test this separately and then name the test properly.
Whichever suits you.

Comment (sanchita) [reply app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: Also I see that the assertions are happening on the position on the patient in the list rather than the `fullName` itself.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt [hunk 4]
```diff
@@ -1932,5 +1932,169 @@ class PatientRepositoryAndroidTest {
...
+
+    assertThat(allPatientsInFacilityB)
+        .containsExactly(patientsInFacilityB[2].patient.uuid)
+        .inOrder()
+
+    assertThat(allPatientsInFacilityC).isEmpty()
+  }
+
+  @Test
+  fun querying_all_patients_in_a_facility_with_bps_in_different_facilities_should_set_the_last_bp_to_the_latest_one() {
```
Comment (sanchita) [inline app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: Rephrase this to `querying_all_patients_in_a_facility_should_set_the_last_bp_to_the_last_recorded_one_within_any_facility`

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt [hunk 5]
```diff
@@ -1932,5 +1932,169 @@ class PatientRepositoryAndroidTest {
...
+
+    fun recordPatientWithBps(
```
Comment (sanchita) [inline app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: This function is doing two separate operations.
Can you split this in two?
One function should create the profile whereas the other could record Bps.

File: app/src/main/java/org/simple/clinic/patient/PatientSearchResult.kt [hunk 6]
```diff
@@ -90,6 +90,18 @@ data class PatientSearchResult(
...
+    @Query("""
+      SELECT * FROM ($mainQuery) AllSearchResults
+      INNER JOIN (
+        SELECT DISTINCT P.uuid FROM Patient P
+          INNER JOIN BloodPressureMeasurement BP ON BP.patientUuid = P.uuid
+          WHERE BP.deletedAt IS NULL AND P.status = :status AND BP.facilityUuid = :facilityUuid
+      ) PatientsAtFacility ON AllSearchResults.uuid = PatientsAtFacility.uuid
+      GROUP BY AllSearchResults.uuid ORDER BY AllSearchResults.fullName
+    """)
+    fun searchInFacilitySortedByName(facilityUuid: UUID, status: PatientStatus): Flowable<List<PatientSearchResult>>
```
Comment (sanchita) [inline app/src/main/java/org/simple/clinic/patient/PatientSearchResult.kt]: Rename this to `searchInFacilityAndSortByName`

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt [hunk 7]
```diff
@@ -1932,5 +1932,169 @@ class PatientRepositoryAndroidTest {
...
+      }
+
+      patientRepository.save(listOf(patientProfile)).blockingAwait()
+      bloodPressureRepository.save(bps).blockingAwait()
+
+      return patientProfile
+    }
+
+    fun allPatientIdsAndLastBpRecordedAtFacility(facility: Facility): List<Pair<UUID, Instant>> {
+      return patientRepository.allPatientsInFacility(facility).blockingFirst().map { it.uuid to it.lastBp!!.takenOn }
```
Comment (sanchita) [inline app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: More than two Rx operators are being chained.
Please format each operator to a separate line.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt [hunk 8]
```diff
@@ -1932,5 +1932,169 @@ class PatientRepositoryAndroidTest {
...
+    assertThat(allPatientsInFacilityC).isEmpty()
+  }
+
+  @Test
+  fun querying_all_patients_in_a_facility_with_bps_in_different_facilities_should_set_the_last_bp_to_the_latest_one() {
+
+    fun recordPatientWithBps(
+        patientUuid: UUID,
+        patientName: String,
+        recordBpsAt: List<Triple<UUID, Facility, Instant>>
```
Comment (sanchita) [inline app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: Using a `Triple` here makes it difficult to understand that what needs to be passed in this.
Can you instead use three function arguments?
This will also become concise when you split the function.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt [hunk 9]
```diff
@@ -1932,5 +1932,169 @@ class PatientRepositoryAndroidTest {
...
+        "Patient B",
+        listOf(
+            Triple(bpUuid1ForPatientB, facilityB, now.plusSeconds(5)),
+            Triple(bpUuid2ForPatientB, facilityA, now.minusSeconds(1)),
+            Triple(bpUuid3ForPatientB, facilityB, now)
+        )
+    )
+
+    // when
+    val allPatientsInFacilityA = allPatientIdsAndLastBpRecordedAtFacility(facilityA)
```
Comment (sanchita) [inline app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: Why do we need to create a `Pair` here?
Again this makes it difficult to read and understand what we want to verify.
In the assertions, can't we directly verify ids and bp recordedAt for `allPatientsInFacility`

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt [hunk 10]
```diff
@@ -1932,5 +1932,169 @@ class PatientRepositoryAndroidTest {
...
+        listOf(
+            Triple(bpUuid1ForPatientA, facilityA, now.plusSeconds(1)),
+            Triple(bpUuid2ForPatientA, facilityB, now.minusSeconds(1)),
+            Triple(bpUuid3ForPatientA, facilityA, now)
+        )
+    )
+
+    val bpUuid1ForPatientB = UUID.fromString("00e39456-f5e4-4538-956d-7e973ec5da88")
+    val bpUuid2ForPatientB = UUID.fromString("b72e28cd-1b46-4c72-84cd-58530c2829e3")
+    val bpUuid3ForPatientB = UUID.fromString("c18b41d5-260e-4a26-8030-45c621ded98d")
```
Comment (sanchita) [inline app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: I dont see any of these generated Ids being used further in the method.
We can leave Ids to be auto-generated.
Same for `Patient A` as well.

File: app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityListItem.kt [hunk 11]
```diff
@@ -43,11 +50,19 @@ sealed class AllPatientsInFacilityListItem {
...
+        holder: ViewHolderX,
+        eventSubject: Subject<Event>
+    ) {
       holder.patientSearchResultView.render(patientSearchResult, facility)
+      holder.itemView.setOnClickListener { eventSubject.onNext(Event.SearchResultClicked(patientSearchResult)) }
     }
   }
 
+  sealed class Event {
+    data class SearchResultClicked(val patientSearchResult: PatientSearchResult): AllPatientsInFacilityListItem.Event()
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityListItem.kt]: @ragunathjawahar and I talked about removing the `UiEvent` interface and having individual event types per feature. This is a little hard to do for other screens at the moment, but we could start doing this for events for new features.

File: app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt [hunk 12]
```diff
@@ -114,4 +100,34 @@ class PatientSearchScreenController @Inject constructor() : ObservableTransforme
         .map { it.patientUuid }
         .map { patientUuid -> { ui: Ui -> ui.openPatientSummary(patientUuid) } }
   }
+
+  private fun toggleAllPatientsVisibility(events: Observable<UiEvent>): ObservableSource<UiChange> {
+    return events
+        .ofType<SearchQueryNameChanged>()
+        .map { it.name.isBlank().not() }
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt]: Good point, I'll do this.

File: app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt [hunk 13]
```diff
@@ -114,4 +100,34 @@ class PatientSearchScreenController @Inject constructor() : ObservableTransforme
...
+              ui.showAllPatientsInFacility()
+            }
+          }
+        }
+  }
+
+  private fun toggleSearchButtonVisibility(events: Observable<UiEvent>): ObservableSource<UiChange> {
+    return events
+        .ofType<SearchQueryNameChanged>()
+        .map { it.name.isBlank().not() }
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt]: Good point, I'll do this.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt (hunk 2 above)
Comment (vinaysshenoy) [reply app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: In a nutshell, we are planning to reduce the amount of variability in our tests. Part of doing this is reducing the randomly generated data in our tests and statically generate constant data in the test class itself. We can explain this in more detail next week.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt (hunk 3 above)
Comment (vinaysshenoy) [reply app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: Good point, I'll look at making this test more understandable.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt (hunk 5 above)
Comment (vinaysshenoy) [reply app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: Sure, I'll do this.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt (hunk 8 above)
Comment (vinaysshenoy) [reply app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: Right. I'll do this when splitting the functions.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt (hunk 9 above)
Comment (vinaysshenoy) [reply app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: We could. I'll look at making this test more understandable.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt (hunk 10 above)
Comment (vinaysshenoy) [reply app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: See above comment about reducing the amount of random data in our tests. I think @ragunathjawahar will be able to explain this decision in a better way than I can.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt (hunk 3 above)
Comment (vinaysshenoy) [reply app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: I tried to make it more clear what is happening. Let's go over this again in a bit and see if it makes better sense now?

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt (hunk 9 above)
Comment (vinaysshenoy) [reply app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt]: Take a look now? I've rewritten the test to make it (hopefully) more understandable.
